### PR TITLE
HHKB YDKB Yang BT Controller

### PR DESCRIPTION
**Pull request by jram0421**
Description

Adds support for the Yang controller for HHKB Pro 2.

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [x] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* 

## Checklist

- [ ] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [ ] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [ ] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: keyboards/hhkb_yang/MEMO.md [hunk 1]
```diff
@@ -0,0 +1,94 @@
+## Hardware Information
+
+The YANG HHKB BLE controller design is similiar to hasu's
+controller. Most pins are compatiable.
```
Comment (Erovia) [inline keyboards/hhkb_yang/MEMO.md:4]:
```suggestion
controller. Most pins are compatible.
```

File: keyboards/hhkb_yang/MEMO.md [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+## How to flash LUFA MassStorage bootloader on Linux
+
+The FAT filesystem on Linux very often cannot flush the write cache,
+leading to broken firmware in the flash.
+
+We can use `dd` to write to the virtual block storage directly to
+bypass the vfs layer.
+
+```
+dd if=FLASH.bin of=/dev/sda seek=4
```
Comment (Erovia) [inline keyboards/hhkb_yang/MEMO.md:45]:
```suggestion

```
This is a nice and easy way to break most users installation.
Given that for most systems `/dev/sda` is the main drive and ppl are prone to blindly running commands from the internet, this is a disaster in the making.

If simply calling `sync` after a `cp` is not an option, I'd rather have a short guide here on finding the appropriate device (with `lsblk` or something) than this command.

File: keyboards/hhkb_yang/rules.mk [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+#   change yes to no to disable
+#
+BOOTMAGIC_ENABLE = lite     # Virtual DIP switch configuration
+MOUSEKEY_ENABLE = yes       # Mouse keys
+EXTRAKEY_ENABLE = yes       # Audio control and System control
+CONSOLE_ENABLE = no         # Console for debug
+COMMAND_ENABLE = yes        # Commands for debug and configuration
+# if this doesn't work, see here: https://github.com/tmk/tmk_keyboard/wiki/FAQ#nkro-doesnt-work
+NKRO_ENABLE = no            # USB Nkey Rollover
+
```
Comment (Erovia) [inline keyboards/hhkb_yang/rules.mk:20]:
```suggestion
LAYOUTS = 60_hhkb
```

File: keyboards/hhkb_yang/rules.mk [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+SRC += matrix.c
+
+# Disable debounce
+DEBOUNCE_TYPE = custom
+
+PROGRAM_CMD = if [ -z "$(LUFA_DEVICE)" ]; then \
+		echo "Abort: set LUFA_DEVICE variable first"; \
+	else \
+		sudo umount $(LUFA_DEVICE) \
+		&& sudo dd if=$(BUILD_DIR)/$(TARGET).bin of=$(LUFA_DEVICE) seek=4 \
```
Comment (Erovia) [inline keyboards/hhkb_yang/rules.mk:34]: I'm okay-ish with this as the user need to set `LUFA_DEVICE`.

Although I'm not sure if it will run if the device is not mounted before flashing.  
Changing `&&` to just `;` or even better a check might be needed.

File: keyboards/hhkb_yang/readme.md [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+
+    make hhkb_yang:default
+
+To flash, first boot your keyboard into bootloader (hold ESC and attach usb cable):
+
+    make LUFA_DEVICE=/dev/sda hhkb_yang:default:flash
+
+LUFA_DEVICE should point to the emulated storage device by the lufa-ms
+bootloader. **Caution**: if set to incorrect device it may wipe out
+your actual disk.
```
Comment (Erovia) [inline keyboards/hhkb_yang/readme.md:22]:
```suggestion
    make LUFA_DEVICE=/dev/sda hhkb_yang:default:flash

LUFA_DEVICE should point to the emulated storage device by the lufa-ms
bootloader. **Caution**: if set to incorrect device it may wipe out
your actual disk.
```

Even though it has a warning text, I'd rather not explicitly mention `/dev/sda`, as most will have that device and in most cases that's not gonna be the one they want to `dd`.

**Review by Erovia — CHANGES_REQUESTED** [review]
Personally, I'd rather have this implementation under `keyboards/hhkb/yang` than in its own directory.

File: tmk_core/protocol/lufa/adafruit_ble_uart.c [hunk 6]
```diff
@@ -0,0 +1,377 @@
+#include "adafruit_ble.h"
```
Comment (drashna) [inline tmk_core/protocol/lufa/adafruit_ble_uart.c:1]: The UART BLE code should be a separate PR, and not part of the keyboard PR. 

It should be separate PRs.

**Comment by kanru** [comment]
Thank you, @jram0421, for opening the pull request.

@Erovia I was thinking about to put the directory under `keyboard/hhkb/yang` too, but I didn't want to deal with sharing code with HHKB ansi. Is it ok that `keyboard/hhkb/yang` is a totally new implementation?

Your other suggestions for flashing is great. I'll update them.

@drashna I opened https://github.com/qmk/qmk_firmware/issues/10674 a while ago to discuss how to structure the BLE code. If can go ahead and separate the bits to a separate PR and have discussion there.

**Comment by Erovia** [comment]
Currently everything under `keyboards/hhkb` are separate implementations.
Imho it still makes sense to keep them in one place, where the users would logically look for them.

File: keyboards/hhkb_yang/rules.mk (hunk 4 above)
Comment (fauxpark) [reply keyboards/hhkb_yang/rules.mk:34]: I would rather not have `PROGRAM_CMD` at all, tbh... I think this would only work in Linux, and maybe macOS.

File: keyboards/hhkb_yang/hhkb_yang.c [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+    setPinOutput(F1);
+    setPinOutput(F4);
+    writePinLow(F0);
+    writePinLow(F1);
+    writePinLow(F4);
+
+    // Turn on switch PCB
+    setPinOutput(D6);
+    writePinLow(D6);
+}
```
Comment (drashna) [inline keyboards/hhkb_yang/hhkb_yang.c:71]:
```suggestion

    keyboard_pre_init_user();
}
```

File: keyboards/hhkb_yang/hhkb_yang.c [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+
+void suspend_power_down_kb(void) {
+    if (power_save_level > 2) {
+        // Disable UART TX to avoid current leakage
+        UCSR1B &= ~_BV(TXEN1);
+        // Power down BLE module
+        writePinHigh(D5);
+    }
+}
```
Comment (drashna) [inline keyboards/hhkb_yang/hhkb_yang.c:80]:
```suggestion

    suspend_power_down_user();
}
```

File: keyboards/hhkb_yang/hhkb_yang.c [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+
+void suspend_wakeup_init_kb(void) {
+    // Power up BLE module
+    writePinLow(D5);
+    // Enable UART TX
+    UCSR1B |= _BV(TXEN1);
+}
```
Comment (drashna) [inline keyboards/hhkb_yang/hhkb_yang.c:87]:
```suggestion

    suspend_wakeup_init_kb();
}
```

File: keyboards/hhkb_yang/hhkb_yang.c [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+
+bool led_update_kb(led_t led_state) {
+    bool res = led_update_user(led_state);
+    if (res) {
+        writePin(F4, led_state.caps_lock);
+    }
+    return res;
```
Comment (drashna) [inline keyboards/hhkb_yang/hhkb_yang.c:94]: Actually, you can remove this and just define the caps lock LED. 

https://docs.qmk.fm/#/feature_led_indicators?id=configuration-options

File: keyboards/hhkb_yang/keymaps/kanru/config.h [hunk 11]
```diff
@@ -0,0 +1,4 @@
+#define MOUSEKEY_DELAY 90
```
Comment (drashna) [inline keyboards/hhkb_yang/keymaps/kanru/config.h:1]: Could you add a GPL2+ compatible license header here? 

For instance: 
https://github.com/qmk/qmk_firmware/blob/28ff51175b6d45ce5a8dbfe7ed6e7a12df3bc8a8/quantum/template/base/keyboard.c#L1-L15

Also, this should have a `#pragma once` include guard before any of the settings.

File: keyboards/hhkb_yang/hhkb_yang.c [hunk 12]
```diff
@@ -0,0 +1,104 @@
+#include "hhkb_yang.h"
```
Comment (drashna) [inline keyboards/hhkb_yang/hhkb_yang.c:1]: Could you add a GPL2+ compatible license header here? 

For instance: 
https://github.com/qmk/qmk_firmware/blob/28ff51175b6d45ce5a8dbfe7ed6e7a12df3bc8a8/quantum/template/base/keyboard.c#L1-L15

File: keyboards/hhkb_yang/hhkb_yang.h [hunk 13]
```diff
@@ -0,0 +1,21 @@
+#pragma once
```
Comment (drashna) [inline keyboards/hhkb_yang/hhkb_yang.h:1]: Could you add a GPL2+ compatible license header here? 

For instance: 
https://github.com/qmk/qmk_firmware/blob/28ff51175b6d45ce5a8dbfe7ed6e7a12df3bc8a8/quantum/template/base/keyboard.c#L1-L15

File: keyboards/hhkb_yang/keymaps/default/keymap.c [hunk 14]
```diff
@@ -0,0 +1,56 @@
+/*  -*-  eval: (turn-on-orgtbl); -*-
```
Comment (drashna) [inline keyboards/hhkb_yang/keymaps/default/keymap.c:1]: Could you add a GPL2+ compatible license header here? 

For instance: 
https://github.com/qmk/qmk_firmware/blob/28ff51175b6d45ce5a8dbfe7ed6e7a12df3bc8a8/quantum/template/base/keyboard.c#L1-L15

File: keyboards/hhkb_yang/keymaps/kanru/keymap.c [hunk 15]
```diff
@@ -0,0 +1,82 @@
+/*  -*-  eval: (turn-on-orgtbl); -*-
```
Comment (drashna) [inline keyboards/hhkb_yang/keymaps/kanru/keymap.c:1]: Could you add a GPL2+ compatible license header here? 

For instance: 
https://github.com/qmk/qmk_firmware/blob/28ff51175b6d45ce5a8dbfe7ed6e7a12df3bc8a8/quantum/template/base/keyboard.c#L1-L15

**Review by drashna — CHANGES_REQUESTED** [review]

File: keyboards/hhkb_yang/MEMO.md [hunk 16]
```diff
@@ -0,0 +1,94 @@
...
+   to save power. Sensing PCB is powered down between scans.
+
+2. Level 2: after idling for longer (MATRIX_POWER_SAVE_TIMEOUT_L2_MS)
+   we entry this state. Matrix scan is skipped until the time lapses
+   900ms.
+
+2. Level 3: sleep mode is activated after a longer timeout
+   (MATRIX_POWER_SAVE_TIMEOUT_L3_MS) Bluetooth module is powered down.
+
+## Batter reading
```
Comment (yoyostile) [inline keyboards/hhkb_yang/MEMO.md:78]:
```suggestion
## Battery reading
```

**Comment by yoyostile** [comment]
I'd love to see this in `master`!

File: keyboards/hhkb_yang/MEMO.md [hunk 17]
```diff
@@ -0,0 +1,94 @@
...
+
+VBAT is connected to AIN6 pin on the MDBT40 module and the AREF pin is
+the reference voltage. Doing a ADC with AT+HWDAC=6 will return the
+difference between VBAT and VREF.
+
+It seems when fully charged the ADC read is 550. Likely VREF is 3311mV
+and the fully charged VBAT is thus 3861mV.
+
+Enable battery service with AT+GLEBATTEN=1 first then we can update the
+battery level by using AT+GLEBATTVAL=%d
```
Comment (saysjonathan) [inline keyboards/hhkb_yang/MEMO.md:88]:
```suggestion
Enable battery service with AT+BLEBATTEN=1 first then we can update the
battery level by using AT+BLEBATTVAL=%d
```

**Comment by kanru** [comment]
Addressed all the comments and extracted the Adafruit UART code to separate PR here https://github.com/qmk/qmk_firmware/pull/11403

**Comment by saysjonathan** [comment]
@kanru The Yang docs mention the ability to [output the battery percentage as text,](http://help.ydkb.io/doku.php?id=en:kb-mods:hhkb-ble#battery_charging) which is bound to `Fn-E` in the offical yang defaults. Is it possible to replicate that functionality in qmk?

**Comment by kanru** [comment]
> @kanru The Yang docs mention the ability to [output the battery percentage as text,](http://help.ydkb.io/doku.php?id=en:kb-mods:hhkb-ble#battery_charging) which is bound to `Fn-E` in the offical yang defaults. Is it possible to replicate that functionality in qmk?

yes, it's totally possible. It's very easy to implement via extra commands or user commands. I have a local change in my keymap does just that. Let me push the commit in a bit.

**Comment by kanru** [comment]
ping?

**Comment by yoyostile** [comment]
@kanru how long does your battery last with qmk? I'm down to like 4-5 days using a 2000mAh battery.

**Comment by kanru** [comment]
> @kanru how long does your battery last with qmk? I'm down to like 4-5 days using a 2000mAh battery.

About 3-4 days (using 8~9 hours per day) using a 1200mAh battery. I think it can still be optimized to double or triple the use time but I don't have a oscilloscope yet so my ability to troubleshoot is limited.

I suspect the current sleep state in qmk is not deep or long enough, after all it was for usb power saving.

File: keyboards/hhkb/yang/matrix.c [hunk 18]
```diff
@@ -0,0 +1,211 @@
+/*
```
Comment (drashna) [inline keyboards/hhkb/yang/matrix.c:1]: Would it be possible to convert this to matrix lite?

Comment (kanru) [reply keyboards/hhkb/yang/matrix.c:1]: Yes. I was not familiar with the code base. Converting to matrix lite makes the code a lot simplier. Thanks!

File: keyboards/hhkb/yang/MEMO.md [hunk 19]
```diff
@@ -0,0 +1,135 @@
+## Hardware Information
```
Comment (drashna) [inline keyboards/hhkb/yang/MEMO.md:1]: Could this file be renamed, so it's lower case?

File: keyboards/hhkb/yang/matrix.c (hunk 18 above)
Comment (drashna) [reply keyboards/hhkb/yang/matrix.c:1]: Not a problem!

File: keyboards/hhkb/yang/MEMO.md (hunk 19 above)
Comment (kanru) [reply keyboards/hhkb/yang/MEMO.md:1]: Sure

**Review by drashna — APPROVED** [review]

**Comment by kanru** [comment]
Bump again

File: keyboards/hhkb/yang/readme.md [hunk 20]
```diff
@@ -0,0 +1,118 @@
+# HHKB Alternate Controller (YANG HHKB BLE Mod)
+
+![YANG HHKB BLE Mod](https://i.imgur.com/aZP1GYc.jpeg)
+
+An alternative controler for the HHKB designed by YANG (yangdigi)
+based on the hasu controller.
+
+Keyboard Maintainer: [Kan-Ru Chen](https://github.com/kanru)  
+Hardware Supported: YANG HHKB BLE Controller  
+Hardware Availability: https://kbdfans.com/products/hhkb-ble-mod-upgrade-module
```
Comment (noroadsleft) [inline keyboards/hhkb/yang/readme.md:10]: Having this section as a list is preferred.

```suggestion
* Keyboard Maintainer: [Kan-Ru Chen](https://github.com/kanru)
* Hardware Supported: YANG HHKB BLE Controller
* Hardware Availability: https://kbdfans.com/products/hhkb-ble-mod-upgrade-module
```

File: keyboards/hhkb/yang/rules.mk [hunk 21]
```diff
@@ -0,0 +1,31 @@
...
+
+# Disable bluetooth until the UART code is merged
+# BLUETOOTH = AdafruitBLE_UART
+
+# Custom matrix file for the HHKB
+CUSTOM_MATRIX = lite
+SRC += matrix.c
+
+# Disable debounce
+DEBOUNCE = 0
```
Comment (noroadsleft) [inline keyboards/hhkb/yang/rules.mk:31]: I believe this needs to be set in `config.h` (as `#define DEBOUNCE 0`) rather than here.

**Review by noroadsleft — CHANGES_REQUESTED** [review]

File: keyboards/hhkb/yang/rules.mk (hunk 21 above)
Comment (drashna) [reply keyboards/hhkb/yang/rules.mk:31]: ^ Correct.  In the rules file, it won't actually do anything. 

Though, leaving it undefined will also set it to 0.

**Comment by floyza** [comment]
Bump, it would be nice to have this merged

**Comment by raystudioco** [comment]
Bump, has this been merged yet? 
@kanru thank you for porting as QMK. I have a question, is there a Key defined for switching BT paired device could be used in keymap?

**Comment by kanru** [comment]
> Bump, has this been merged yet? @kanru thank you for porting as QMK. I have a question, is there a Key defined for switching BT paired device could be used in keymap?

Extra command is defined to force a BT repair by `LSHIFT + RSHIFT + S`.

You can also define keybindings in your keymap.

For example:

```
bool process_record_user(uint16_t keycode, keyrecord_t *record) {
    switch (keycode) {
#ifdef BLUETOOTH_ENABLE
        case KC_BTS:
            if (record->event.pressed) {
                adafruit_ble_reconnect();
            }
            return false;
#endif
        default:
            return true;
    }
}
```

It was limited by the Adafruit firmware on the BT module, which only supports one bonded device. So you'll need to reconnect until the correct device is selected.

---
TBH I'm not sure when this will be merged, it's been so long. Feel free to use or submit PR to my fork.

Due to the limitation of the Adafruit firmware mentioned above and wanting to have a more open hardware I have switched to a hand wired nice!nano board and custom ZMK build. I'm very happy with the new setup. Both the board and the firmware are fully open source, more power efficient (up to 15 days usage with a 1200 mAh battery), ZMK is a real RTOS and supports virtually unlimited bonded devices. I plan to also release the new setup this year.

**Comment by peepeetee** [comment]
I'd recommend reopening this PR, if you're still up to it.

Also, I've recently created a PR to collect known YDKB boards in a directory #15681 , you might want to check that out if you decide to reopen this PR
